plugins/analysis/common/src/main/kotlin/arrow/meta/plugins/analysis/phases/analysis/solver/check/Expressions.kt

```diff
@@ -1373,22 +1402,24 @@ private fun SolverState.checkWhileExpression(
     }
 }
 
-private fun SolverState.checkLoopBody(body: Expression?, data: CheckData): ContSeq<StateAfter> {
-  return checkExpressionConstraintsWithNewName("loop", body, data).map { stateAfter ->
-    // only keep working on this branch
-    // if we had a 'return' inside
-    // otherwise the other branch is enough
-    // if we decide to abort we need to 'pop',
-    // because the one from 'bracket' won't run
-    when (stateAfter.returnInfo) {
-      is ExplicitLoopReturn -> {
-        prover.pop()
-        abort()
-      }
-      is ExplicitBlockReturn -> stateAfter
-      else -> {
-        prover.pop()
-        abort()
+private fun SolverState.checkLoopBody(
+  body: Expression?,
+  afterBody: List<Expression>,
+  data: CheckData
+): ContSeq<StateAfter> {
+  return checkExpressionConstraintsWithNewName("loop", body, data).flatMap { stateAfter ->
+    // check the additional updates (used when describing "three-piece" for loops)
+    val afterBodyVar = solver.makeObjectVariable(newName(data.context, "afterBody", body))
+    checkBlockExpression(afterBodyVar, afterBody, false, stateAfter.data).map {
+      // only keep working on this branch
+      // if we had a 'return' inside
+      // otherwise the other branch is enough
+      // if we decide to abort we need to 'pop',
+      // because the one from 'bracket' won't run
+      when (stateAfter.returnInfo) {
+        is ExplicitLoopReturn -> abort()
+        is ExplicitBlockReturn -> stateAfter
+        else -> abort()
       }
```

Comment:
What is else here? Can we make it exhaustive?

Reply:
Fixed